### PR TITLE
nixos: systemd: split module up into multiple files

### PR DESCRIPTION
**Pull request by bobvanderlinden**
###### Description of changes

Currently the systemd module is quite large. It contains multiple components that could be split off into separate files.

The idea behind this refactor is to not only make the modules more readable, but it is also a step towards more reusable systemd configuration building.

This PR does introduce a new option called `systemd.additionalUpstreamUserUnits`. It was introduced to have systemd-tmpfiles specify its own user units, instead of systemd-user doing that for systemd-tmpfiles. I think the way these upstream units still needs a bit of work, but I intend to do that in a separate PR.

A next step is to generalize unit directory building between systemd system and systemd user. The main intention is to also reuse this for systemd in initrd (https://github.com/NixOS/nixpkgs/pull/120015). It avoids creating yet another system of unit building.

I did not ran all tests, but I did ran the systemd tests as well as applied this configuration to my own system.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/module-list.nix [hunk 1]
```diff
@@ -1163,7 +1163,12 @@
   ./system/boot/stage-1.nix
   ./system/boot/stage-2.nix
   ./system/boot/systemd.nix
+  ./system/boot/systemd-coredump.nix
```
Comment (Artturin) [inline nixos/modules/module-list.nix:1166]:
```suggestion
  ./system/boot/systemd/coredump.nix
```

Imo would be nicer like this

File: nixos/modules/system/boot/systemd-user.nix [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      "session.slice"
+      "shutdown.target"
+      "smartcard.target"
+      "sockets.target"
+      "sound.target"
+      "systemd-exit.service"
+      "timers.target"
+      "xdg-desktop-autostart.target"
+    ]
+    ++ config.systemd.additionalUpstreamUserUnits;
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/systemd-user.nix:51]: That isn't the formatting that was used before. Lets keep that and only split things

https://github.com/NixOS/nixpkgs/pull/164016/files#diff-5ea4693beb4bdd8f7efcdd3204ceea67c86f59b3d1c28cde063ddf03e1d20fbfL188-L210

File: nixos/modules/system/boot/systemd-journald.nix [hunk 3]
```diff
@@ -0,0 +1,143 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/systemd-journald.nix:6]:
```suggestion
{ config, lib, pkgs, ... }:

```
I would suggest to keep the style of the formatting like it was before. This is also nixpkgs-fmt conform. win win.

File: nixos/modules/module-list.nix (hunk 1 above)
Comment (cole-h) [reply nixos/modules/module-list.nix:1166]: I agree -- if we're splitting systemd into related modules, they should all go under an umbrella `systemd` directory.

File: nixos/modules/system/boot/systemd-user.nix (hunk 2 above)
Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd-user.nix]: I redid the commits to align them with the original formatting again :+1:

File: nixos/modules/module-list.nix (hunk 1 above)
Comment (bobvanderlinden) [reply nixos/modules/module-list.nix]: I like this one as well. My previous approach aligned with `systemd-nspawn`. I have moved `systemd-nspawn` to the `systemd` directory as well in a separate commit. :+1:

File: nixos/modules/system/boot/systemd-journald.nix (hunk 3 above)
Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd-journald.nix]: Done :+1:

**Comment by bobvanderlinden** [comment]
Unfortunately I couldn't make the build unchanged, because the order of `upstreamSystemUnits` and `upstreamUserUnits` changed (due to merging the lists). The order of those lists shouldn't affect functionality.

I ran for each of the commits to make sure they succeed:

```
nix-build nixos/release.nix -A tests.systemd.x86_64-linux
```

**Comment by cole-h** [comment]
cc @NixOS/systemd

File: nixos/modules/system/boot/systemd/user.nix [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      type = with types; attrsOf (submodule [ { options = timerOptions; } unitConfig ] );
+      description = "Definition of systemd per-user timer units.";
+    };
+
+    systemd.additionalUpstreamUserUnits = mkOption {
+      default = [];
+      type = types.listOf types.str;
+      example = [];
+      description = ''
+        Additional user units shipped with systemd that shall be enabled.
```
Comment (flokli) [inline nixos/modules/system/boot/systemd/user.nix:110]: For which user(s) will these be enabled? All users that have a login session?

If we trace its usages, I can see they should end up as files in `/etc/systemd/user`, alongside with all the ones in the static `upstreamUserUnits` list, but it probably shouldn't hurt to write this down, and refer to the appropriate systemd manpage.

Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd/user.nix:110]: The units will just be available to `systemd --user`. This works the same as before. I doubt any NixOS user would want to add/remove someting to `upstreamUserUnits`. Unless they are developing NixOS itself. Same goes for `upstreamSystemUnits`.

I'm thinking it might be best to add `internal = true;` and possibly `visible = false;` to this option.

Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd/user.nix]: I've opted for expanding on the description a bit more to mention per-user systemd instances and also make the option internal.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd/user.nix]: @flokli I think that's tangential to this PR, given this PR is just moving around existing code.

**Comment by ElvishJerricco** [comment]
I would really like to see this merge. It's been helpful in my systemd-initramfs effort

**Comment by arianvp** [comment]
@GrahamcOfBorg test systemd

**Comment by arianvp** [comment]
Could we squash the `systemd-user` bits into a single commit or split it off into a new PR?  Given it adds new functionaltiy it's a bit hard to follow that it's interwoven with plain refactorings.

Apart from that LGTM

**Review by arianvp — APPROVED** [review]

File: nixos/modules/system/boot/systemd/coredump.nix [hunk 5]
```diff
@@ -0,0 +1,58 @@
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+
+let
+  cfg = config.systemd.coredump;
+  systemd = config.systemd.package;
+in {
+  options = {
+    systemd.coredump.enable = mkOption {
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/coredump.nix:10]:
```suggestion
  options.systemd.coredump = {
    enable = mkOption {
```

File: nixos/modules/system/boot/systemd/coredump.nix [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      default = true;
+      type = types.bool;
+      description = ''
+        Whether core dumps should be processed by
+        <command>systemd-coredump</command>. If disabled, core dumps
+        appear in the current directory of the crashing process.
+      '';
+    };
+
+    systemd.coredump.extraConfig = mkOption {
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/coredump.nix:20]:
```suggestion
    extraConfig = mkOption {
```

File: nixos/modules/system/boot/systemd/coredump.nix [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      description = ''
+        Extra config options for systemd-coredump. See coredump.conf(5) man page
+        for available options.
+      '';
+    };
+  };
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Coredumps.
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/coredump.nix:33]:
```suggestion

```

File: nixos/modules/system/boot/systemd/coredump.nix (hunk 5 above)
Comment (dasJ) [reply nixos/modules/system/boot/systemd/coredump.nix:10]: Debatable change but if done should also done to the other modules

File: nixos/modules/system/boot/systemd/journald.nix [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      type = types.bool;
+      description = ''
+        Whether to forward log messages to syslog.
+      '';
+    };
+  };
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Journal.
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/journald.nix:85]:
```suggestion

```

File: nixos/modules/system/boot/systemd/logind.nix [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        Specifies what to do when the laptop lid is closed and the system is
+        on external power. By default use the same action as specified in
+        services.logind.lidSwitch.
+      '';
+    };
+  };
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Login stuff.
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/logind.nix:81]:
```suggestion

```

File: nixos/modules/system/boot/systemd/tmpfiles.nix [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        If a <filename>lib</filename> output is available, rules are searched there and only there.
+        If there is no <filename>lib</filename> output it will fall back to <filename>out</filename>
+        and if that does not exist either, the default output will be used.
+      '';
+    };
+  };
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Temporary file creation / cleanup.
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/tmpfiles.nix:45]:
```suggestion

```

**Review by dasJ — COMMENTED** [review]
Changes look great, apart from my (somewhat opinionated) minor style nits ;)

I'd be happy if this went through without my concerns being adressed, so I'm only commenting here and not requesting changes.

File: nixos/modules/system/boot/systemd/coredump.nix (hunk 5 above)
Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd/coredump.nix:10]: Meh, most files followed the `options = {` style approach, so I followed by example. It's fine to change them all and I'm fine to do that, but it seems more fitting in a separate PR.

If there is an unwritten rule that `options.* =` style approach should be taken, let's write that down somewhere. (linting rules?)

File: nixos/modules/system/boot/systemd/coredump.nix (hunk 6 above)
Comment (bobvanderlinden) [reply nixos/modules/system/boot/systemd/coredump.nix:20]: Same as above. This conflicts with @SuperSandro2000's comment to keep the code as-is in this PR as much as possible to help the review process.

**Comment by bobvanderlinden** [comment]
> I would really like to see this merge. It's been helpful in my systemd-initramfs effort

Review approval would help :smile: :+1:

**Review by dasJ — APPROVED** [review]
LGTM. If nobody brings up arguments why I shouldn't, I'd merge this tomorrow or so

File: nixos/modules/system/boot/systemd/user.nix [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+      type = with types; attrsOf (submodule [ { options = timerOptions; } unitConfig ] );
+      description = "Definition of systemd per-user timer units.";
+    };
+
+    systemd.additionalUpstreamUserUnits = mkOption {
+      default = [];
+      type = types.listOf types.str;
+      example = [];
+      description = ''
+        Additional units shipped with systemd that shall be enabled for per-user systemd instances.
```
Comment (mweinelt) [inline nixos/modules/system/boot/systemd/user.nix:110]:
```suggestion
        Additional units shipped with systemd that should be enabled for per-user systemd instances.
```

I think should is more fitting, since you're not proposing anything here.

**Review by mweinelt — APPROVED** [review]
Looks sane enough. Tested on my desktop and I didn't notice any regression.

**Review by cole-h — APPROVED** [review]

**Review by flokli — APPROVED** [review]

**Comment by bobvanderlinden** [comment]
🥳 thank you all!
